kerastuner/tuners/bayesian.py

```diff
@@ -0,0 +1,287 @@
+import json
+import random
+
+import numpy as np
+from scipy import optimize as scipy_optimize
+from sklearn import gaussian_process
+
+from ..abstractions.tensorflow import TENSORFLOW_UTILS as tf_utils
+from ..engine import hyperparameters as hp_module
+from ..engine import oracle as oracle_module
+from ..engine import tuner as tuner_module
+
+
+class BayesianOptimizationOracle(oracle_module.Oracle):
+    """Bayesian optimization oracle.
+
+    It uses Bayesian optimization with a underlying Gaussian process model.
+    The acquisition function used is upper confidence bound (UCB), which can
+    be found in the following link:
+    https://www.cse.wustl.edu/~garnett/cse515t/spring_2015/files/lecture_notes/12.pdf
+
+    Attributes:
+        num_initial_points: Int. The number of randomly generated samples as initial
+            training data for Bayesian optimization.
+        alpha: Float. Value added to the diagonal of the kernel matrix
+            during fitting. It represents the expected amount of noise
+            in the observed performances in Bayesian optimization.
+        beta: Float. The balancing factor of exploration and exploitation.
+            The larger it is, the more explorative it is.
+        seed: Int. Random seed.
+
+    """
+
+    def __init__(self,
+                 num_initial_points=2,
+                 alpha=1e-4,
+                 beta=2.6,
+                 seed=None):
+        super(BayesianOptimizationOracle, self).__init__()
+        self.num_initial_points = num_initial_points
+        self.alpha = alpha
+        self.beta = beta
+        self.seed = seed or random.randint(1, 1e4)
+        self._seed_state = self.seed
+        self._tried_so_far = set()
+        self._max_collisions = 20
+        self._num_trials = 0
+        self._score = {}
+        self._values = {}
+        self._x = None
+        self._y = None
+        self.gpr = gaussian_process.GaussianProcessRegressor(
+            kernel=gaussian_process.kernels.ConstantKernel(1.0),
+            alpha=self.alpha)
+
+    def populate_space(self, trial_id, space):
+        self.update_space(space)
+        # Generate enough samples before training Gaussian process.
+        if self._num_trials < self.num_initial_points or len(self._score) < 2:
+            self._num_trials += 1
+            values = self._new_trial()
+            self._values[trial_id] = values
+            return {'status': 'RUN', 'values': values}
+        values = self._to_hp_dict(self._generate_vector())
+        self._values[trial_id] = values
+        return {'status': 'RUN', 'values': values}
+
+    def report_status(self, trial_id, status):
+        # TODO
+        raise NotImplementedError
+
+    def save(self, fname):
+        state = {
+            'init_samples': self.num_initial_points,
```

Comment:
This is inconsistent, it should be called `num_initial_points` in the config as well